packages/Button/src/Button.tsx

```diff
@@ -267,61 +292,124 @@ const baseStyle = css`
   }
 `;
 
-export default class Button extends Component {
-  static displayName = 'Button';
+interface SharedButtonProps {
+  variant: Variant;
+  size: Size;
+  className?: string;
+  children?: React.ReactNode;
+  disabled?: boolean;
+}
 
-  static propTypes = {
-    variant: PropTypes.oneOf(['default', 'primary', 'info', 'danger', 'dark']),
-    size: PropTypes.oneOf(['xsmall', 'small', 'normal', 'large']),
-    className: PropTypes.string,
-    children: PropTypes.node,
-    disabled: PropTypes.bool,
-    as: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
-    href: PropTypes.string,
-  };
+interface LinkButtonProps extends HTMLElementProps<'a'>, SharedButtonProps {
+  href: string;
+}
+
+interface ButtonButtonProps
+  extends HTMLElementProps<'button'>,
+    SharedButtonProps {
+  href?: null;
+}
+
+type CustomElementButtonProps = SharedButtonProps & {
+  as: React.ElementType<any>;
+  [key: string]: any;
+};
+
+type ButtonProps =
+  | LinkButtonProps
+  | ButtonButtonProps
+  | CustomElementButtonProps;
+
+function hasCustomRoot(props: ButtonProps): props is CustomElementButtonProps {
+  return (props as any).as != null;
+}
+
+function hasLinkRoot(
+  props: LinkButtonProps | ButtonButtonProps,
+): props is LinkButtonProps {
+  return props.href != null;
+}
+
+const omitProps = [
+  'as',
+  'className',
+  'disabled',
+  'size',
+  'variant',
+  'children',
+] as const;
+
+export default class Button extends Component<ButtonProps> {
+  static displayName = 'Button';
 
   static defaultProps = {
-    variant: 'default',
-    size: 'normal',
+    variant: Variant.Default,
+    size: Size.Normal,
     className: '',
     children: null,
     disabled: false,
-    type: 'button',
   };
 
   render() {
-    const {
-      className,
-      children,
+    const { className, children, disabled, variant, size } = this.props;
+
+    const commonProps = {
+      className: cx(
+        baseStyle,
+        buttonSizes[size],
+        buttonVariants[variant],
+        className,
+      ),
       disabled,
-      variant,
-      size,
-      as,
-      href,
-      ...rest
-    } = this.props;
+      'aria-disabled': disabled,
+    };
 
-    let Root = href ? 'a' : 'button';
+    if (hasCustomRoot(this.props)) {
+      const Root = this.props.as;
+      const rest = omit(this.props, omitProps);
```

Comment:
I might be missing something... how come we can't just use object `rest` syntax?

Reply:
If I try to pull off the `as` prop while destructuring, it will complain because `as` doesn't exist on all of the union members that make up `ButtonProps`. So that means if I use object rest syntax with the destructuring, `rest` is going to contain `as`. This was the best pattern I could come up with to get around that 🤷‍♂

Reply:
Not sure this is better than doing something like this prior to returning the JSX:
```
delete rest.as;
```

Thoughts @hswolff?